src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,214 @@
+                s_initialized = true;
+            }
+            lock (s_registrations)
+            {
+                if (!s_registrations.TryGetValue(_signo, out List<WeakReference<PosixSignalRegistration>>? signalRegistrations))
+                {
+                    signalRegistrations = new List<WeakReference<PosixSignalRegistration>>();
+                    s_registrations.Add(_signo, signalRegistrations);
+                }
+                if (signalRegistrations.Count == 0)
+                {
+                    Interop.Sys.EnablePosixSignalHandling(_signo);
+                }
+                signalRegistrations.Add(new WeakReference<PosixSignalRegistration>(this));
+            }
+            _registered = true;
+        }
+
+        private void Handle(PosixSignalContext context)
+        {
+            lock (_gate)
+            {
+                if (_registered)
+                {
+                    _handler(context);
+                }
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static int OnPosixSignal(int signo, PosixSignal signal)
+        {
+            lock (s_registrations)
+            {
+                if (s_registrations.TryGetValue(signo, out List<WeakReference<PosixSignalRegistration>>? signalRegistrations))
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        var registrations = new PosixSignalRegistration?[signalRegistrations.Count];
+                        bool hasRegistrations = false;
+                        for (int i = 0; i < signalRegistrations.Count; i++)
+                        {
+                            if (signalRegistrations[i].TryGetTarget(out PosixSignalRegistration? registration))
+                            {
+                                registrations[i] = registration;
+                                hasRegistrations = true;
+                            }
+                            else
+                            {
+                                // WeakReference no longer holds an object. PosixSignalRegistration got finalized.
+                                signalRegistrations.RemoveAt(i);
+                                i--;
+                            }
+                        }
+                        if (hasRegistrations)
+                        {
+                            // This is called on the native signal handling thread. We need to move to another thread so
+                            // signal handling is not blocked. Otherwise we may get deadlocked when the handler depends
+                            // on work triggered from the signal handling thread.
+
+                            // For terminate/interrupt signals we use a dedicated Thread
+                            // in case the ThreadPool is saturated.
+                            bool useDedicatedThread = signal == PosixSignal.SIGINT ||
```

Comment:
If we're going to do this, we should probably do it for `SIGHUP`, `SIGTSTP`, and `SIGCONT` as well. I think there's a reasonable argument that from a user's perspective, these are just as urgent as termination signals.

But I still think it would make more sense for each `PosixSignalRegistration` to have a property indicating whether it considers handling the signal urgent. This code could then check whether any of the objects in `registrations` have it set, and if so, spawn a new thread.

Reply:
> If we're going to do this,

This maintains the existing behavior for SIGINT/SIGQUIT which already spawn a thread for the Console event.

> we should probably do it for SIGHUP, SIGTSTP, and SIGCONT as well. I think there's a reasonable argument that from a user's perspective, these are just as urgent as termination signals.

It makes sense to spawn a new thread for SIGINT (and SIGTERM) as those may occur due to the ThreadPool being saturated.

I think this is less the case for SIGHUP, SIGTSTP, and SIGCONT. The ThreadPool should not be saturated, and running them on the ThreadPool is going to be faster than spawning a new Thread.

@alexrp do you want a dedicated thread for these signals?

> But I still think it would make more sense for each PosixSignalRegistration to have a property indicating whether it considers handling the signal urgent. This code could then check whether any of the objects in registrations have it set, and if so, spawn a new thread.

You can create an API proposal.

Reply:
> This maintains the existing behavior for SIGINT/SIGQUIT which already spawn a thread for the Console event.

I don't disagree with that part. 🙂 I noticed you added a new signal to the list, that's why I commented.

> It makes sense to spawn a new thread for SIGINT (and SIGTERM) as those may occur due to the ThreadPool being saturated.

Presumably in response to the user trying to kill the program on the assumption that it's misbehaving?

> I think this is less the case for SIGHUP, SIGTSTP, and SIGCONT. The ThreadPool should not be saturated, and running them on the ThreadPool is going to be faster than spawning a new Thread.

My concern is with TUI-style apps. Here, many of these signals have to do with UI rendering and interaction. As in any UI framework, you want a dedicated thread to handle UI stuff so the UI doesn't feel sluggish (or worse, is rendered completely broken) if lots of work is going on in the background.

That's why I have a single dedicated thread for signal handling in System.Terminal - it sticks around for the lifetime of the program and handles all relevant signals (currently `SIGINT`, `SIGQUIT`, `SIGWINCH`, `SIGCONT`). Perhaps that could be a reasonable compromise? No new API surface for indicating urgency but also no special-casing based on signal code. And maybe the thread could be spawned lazily based on usage of the API.

But if you think this is too much work for too little benefit, I understand. I'm not the one doing the implementation work. 😄

Reply:
> Presumably in response to the user trying to kill the program on the assumption that it's misbehaving?

Yes.

> But if you think this is too much work for too little benefit, I understand. I'm not the one doing the implementation work. smile

I think the current behavior should be good enough. If it turns out not to be the case, we can see what is needed and improve the implementation.